libraries/AP_HAL/RCOutput.cpp

```diff
@@ -0,0 +1,109 @@
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "RCOutput.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <SRV_Channel/SRV_Channel.h>
+
+// SBUS1 constant definitions
+// pulse widths measured using FrSky Sbus/PWM converter
+#define SBUS_BSIZE    25
+#define SBUS_CHANNELS 16
+#define SBUS_MIN 880.0f
+#define SBUS_MAX 2156.0f
+#define SBUS_SCALE (2048.0f / (SBUS_MAX - SBUS_MIN))
+
+/*
+ * build and send sbus1 frame representing first 16 servo channels
+ * input arg is pointer to uart
+ */
+void
+AP_HAL::RCOutput::sbus1_out(uint16_t *pwidth, uint8_t nchan)
+{
+    if (sbus1_uart == nullptr) return;
+
+    // constrain output rate using sbus_frame_interval
+    static uint32_t last_micros = 0;
+    uint32_t now = AP_HAL::micros();
+    if ((now - last_micros) > sbus_frame_interval) {
+        last_micros = now;
+        uint8_t buffer[SBUS_BSIZE] = { 0x0f };  // first byte is always 0x0f
+        uint8_t index = 1;
+        uint8_t offset = 0;
+        uint16_t value;
+
+        /* construct sbus frame representing channels 1 through 16 (max) */
+        if (nchan > SBUS_CHANNELS) nchan = SBUS_CHANNELS;
+        for (unsigned i = 0; i < nchan; ++i) {
+            value = (uint16_t)((pwidth[i] - SBUS_MIN) * SBUS_SCALE);
+
+            /*protect from out of bounds values and limit to 11 bits*/
+            if (value > 0x07ff) {
+                value = 0x07ff;
+            }
+
+            while (offset >= 8) {
+                ++index;
+                offset -= 8;
+            }
+
+            buffer[index] |= (value << (offset)) & 0xff;
+            buffer[index + 1] |= (value >> (8 - offset)) & 0xff;
+            buffer[index + 2] |= (value >> (16 - offset)) & 0xff;
+            offset += 11;
+        }
+        sbus1_uart->write(buffer, sizeof(buffer));
+    }
+}
+
+void AP_HAL::RCOutput::enable_sbus_out(UARTDriver* uart, uint16_t rate) {
+    sbus1_uart = uart;
+    sbus1_uart->begin(100000);
+    sbus1_uart->configure_parity(2);    // enable even parity
+    sbus1_uart->set_stop_bits(2);
+    sbus1_uart->set_unbuffered_writes(true);
+    // subtract 500msec from requested frame interval to allow for latency
+    sbus_frame_interval = (1000UL * 1000UL) / rate - 500;
+    // at 100,000 bps, a 300 bit sbus frame takes 3msec to transfer
+    // require a minimum 700usec interframe gap
+    if (sbus_frame_interval < 3700) {
+        sbus_frame_interval = 3700;
```

Comment:
is the frame interval always 3700? If not maybe a const would be more readable.

Reply:
yes, sbus out is always 100,000 bps